src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,259 @@
+<?php
+
+class WP_HTML_Processor extends WP_HTML_Tag_Processor {
+	public $fully_supported_input = null;
+	public $open_elements = array();
+
+	public function ensure_support() {
+		if ( null !== $this->fully_supported_input ) {
+			return $this->fully_supported_input;
+		}
+
+		$stack = array();
+
+		$p = new WP_HTML_Tag_Processor( $this->html );
```

Comment:
Smart! I kind of don't like that it throws away the progress we have made so far, though

Reply:
what do you mean? my vision for this is that it can grow with our added support.

as for running a first-pass through the document I'm not claiming this is certainly needed. I'm sure it's possible to work it in alongside the normal traversal.

for now though it helped quite a bit, and I figured the cost is meager given that it's a read-only scan.